Firestore/Example/Tests/Integration/API/FIRQueryTests.mm

```diff
@@ -537,6 +537,31 @@ - (void)testQueriesCanUseNotEqualFiltersWithDocIds {
                         (@[ testDocs[@"ab"], testDocs[@"ba"], testDocs[@"bb"] ]));
 }
 
+- (void)testSDKUsesNotEqualFiltersSameAsServer {
+  NSDictionary *testDocs = @{
+    @"a" : @{@"zip" : @(NAN)},
+    @"b" : @{@"zip" : @91102},
+    @"c" : @{@"zip" : @98101},
+    @"d" : @{@"zip" : @"98101"},
+    @"e" : @{@"zip" : @[ @98101 ]},
+    @"f" : @{@"zip" : @[ @98101, @98102 ]},
+    @"g" : @{@"zip" : @[ @"98101", @{@"zip" : @98101} ]},
+    @"h" : @{@"zip" : @{@"code" : @500}},
+    @"i" : @{@"zip" : [NSNull null]},
+    @"j" : @{@"code" : @500},
+  };
+  FIRCollectionReference *collection = [self collectionRefWithDocuments:testDocs];
+
+  // populate cache with all documents first to ensure getDocsFromCache() scans all docs
+  [self readDocumentSetForRef:collection];
+
+  [self checkOnlineAndOfflineQuery:[collection queryWhereField:@"zip" isNotEqualTo:@98101]
+                     matchesResult:@[ @"a", @"b", @"d", @"e", @"f", @"g", @"h" ]];
+
+  [self checkOnlineAndOfflineQuery:[collection queryWhereField:@"zip" isNotEqualTo:@(NAN)]
+                     matchesResult:@[ @"b", @"c", @"d", @"e", @"f", @"g", @"h" ]];
```

Comment:
Please add [this](https://github.com/firebase/firebase-android-sdk/pull/6859#discussion_r2038221788) test case.